File: WinForms/VB/GridMainDemo/Modules/ExcelFiltering.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.Data.Filtering
Imports DevExpress.DXperience.Demos
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid

Namespace DevExpress.XtraGrid.Demos

    Public Partial Class ExcelFiltering
        Inherits TutorialControl

        Public Overrides ReadOnly Property ExportView As BaseView
            Get
                Return bandedGridView1
            End Get
        End Property

        Protected Overrides Sub InitXMLData(ByVal dataFileName As String)
            VehiclesData.InitOrdersData(dataFileName, gridControl1, 10000, 400)
            VehiclesData.InitColumnViewEditors(bandedGridView1)
        End Sub

        Public Overrides ReadOnly Property ShowViewOptions As Boolean
            Get
                Return True
            End Get
        End Property

        Public Sub New()
            CreateWaitDialog()
            InitializeComponent()
            InitVehiclesData()
            SetToolTipController(Me, gridControl1)
            bandedGridView1.HtmlImages = DemoHelper.GetTransmissionGlyphs()
            If Not MainFormHelper.TakeScreens Then
                bandedGridView1.ActiveFilterString = "InRange([ModelPrice], 25000, 80000) And (IsOutlookIntervalToday([SalesDate]) Or IsOutlookIntervalYesterday([SalesDate]))"
                bandedGridView1.MRUFiltersInfo.Add(New ViewFilter(bandedGridView1, CriteriaOperator.Parse("InRange([ModelPrice], 50000, 150000)")))
                bandedGridView1.MRUFiltersInfo.Add(New ViewFilter(bandedGridView1, CriteriaOperator.Parse("IsOutlookIntervalToday([SalesDate]) Or IsOutlookIntervalYesterday([SalesDate])")))
                bandedGridView1.MRUFiltersInfo.Add(New ViewFilter(bandedGridView1, CriteriaOperator.Parse("[SalesDate] > AddDays(LocalDateTimeToday(),-25)")))
            Else
                bandedGridView1.ActiveFilterString = "([ModelPrice] >= 25000 And [ModelPrice] <= 80000)"
            End If
        End Sub

        Protected Overrides ReadOnly Property WhatsThisCodeFileNames As String()
            Get
                Return New String() {"ExcelFiltering"}
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisXMLFileName As String
            Get
                Return "ExcelFiltering"
            End Get
        End Property

        ' Addding custom Filters and Data Items customization
        '<gridControl1>
        Private Sub gridView_FilterPopupExcelData(ByVal sender As Object, ByVal e As FilterPopupExcelDataEventArgs)
            Dim fieldName As String = e.Column.FieldName
            If e.Column Is bcModification Then
                ' Filter Items
                e.AddFilter("<image=A><nbsp>Automatic Transmission (6-speed)", "Contains([" & fieldName & "], '6A')", True)
                e.AddFilter("<image=A><nbsp>Automatic Transmission (8-speed)", "Contains([" & fieldName & "], '8A')", True)
                e.AddFilter("<image=M><nbsp>Manual Transmission (6-speed)", "Contains([" & fieldName & "], '6M')", True)
                e.AddFilter("<image=M><nbsp>Manual Transmission (7-speed)", "Contains([" & fieldName & "], '7M')", True)
                e.AddFilter("<image=V><nbsp>Variomatic Transmission", "Contains([" & fieldName & "], 'VA')", True)
                e.AddFilter("<b>Limited Edition</b>", "Contains([" & fieldName & "], 'Limited')", True)
                ' Data Items
                For Each item In e.DataItems
                    If item.Text.Contains("V6") Then item.HtmlText = item.Text.Replace("V6", "<b>V6</b>")
                    If item.Text.Contains("V8") Then item.HtmlText = item.Text.Replace("V8", "<b>V8</b>")
                    If item.Text.Contains("Limited") Then item.HtmlText = "<image=Ltd><nbsp>" & item.Text
                Next
            End If

            If e.Column Is bcMPGCity Then ' 12-28
                e.AddFilter("Fuel Economy (<color=@Information>High</color>)", "[" & fieldName & "]>25", True)
                e.AddFilter("Fuel Economy (<color=@Warning>Medium</color>)", "[" & fieldName & "]>=15 AND [" & fieldName & "]<=25", True)
                e.AddFilter("Fuel Economy (<color=@Critical>Low</color>)", "[" & fieldName & "]<15", True)
            End If

            If e.Column Is bcMPGHighway Then ' 15-36
                e.AddFilter("Fuel Economy (<color=@Information>High</color>)", "[" & fieldName & "]>20", True)
                e.AddFilter("Fuel Economy (<color=@Warning>Medium</color>)", "[" & fieldName & "]>=20 AND [" & fieldName & "]<=30", True)
                e.AddFilter("Fuel Economy (<color=@Critical>Low</color>)", "[" & fieldName & "]>20", True)
            End If

            ' sales from newer to older by years
            If e.Column Is bcSalesDate Then
                ' using low-level API for inplace values and texts reordering
                Array.Sort(e.Values, e.DisplayTexts, YearsFromNewerToOlder.Comparer)
            End If
        End Sub

        '</gridControl1>
#Region "GenerateReport"
        Public Overrides ReadOnly Property AllowGenerateReport As Boolean
            Get
                Return False
            End Get
        End Property

#End Region
        Private NotInheritable Class YearsFromNewerToOlder
            Implements IComparer(Of Object)

            Public Shared Comparer As IComparer(Of Object) = New YearsFromNewerToOlder()

            Private Sub New()
            End Sub

            '
            Private Function Compare(ByVal x As Object, ByVal y As Object) As Integer Implements IComparer(Of Object).Compare
                Dim dateX = CDate(x)
                Dim dateY = CDate(y)
                Dim yearsResultFromNewerToOlder = dateY.Year.CompareTo(dateX.Year)
                If yearsResultFromNewerToOlder = 0 Then Return dateX.CompareTo(dateY)
                Return yearsResultFromNewerToOlder
            End Function
        End Class
    End Class
End Namespace

File: WinForms/CS/GridMainDemo/Modules/ExcelFiltering.cs
using System;
using System.Collections.Generic;
using DevExpress.Data.Filtering;
using DevExpress.DXperience.Demos;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.XtraGrid.Demos {
    public partial class ExcelFiltering : TutorialControl {
        public override BaseView ExportView {
            get { return bandedGridView1; }
        }
        protected override void InitXMLData(string dataFileName) {
            VehiclesData.InitOrdersData(dataFileName, gridControl1, 10000, 400);
            VehiclesData.InitColumnViewEditors(bandedGridView1);
        }
        public override bool ShowViewOptions { get { return true; } }
        public ExcelFiltering() {
            CreateWaitDialog();
            InitializeComponent();
            InitVehiclesData();
            LicenseInfo.SetToolTipController(this, gridControl1);
            bandedGridView1.HtmlImages = DemoHelper.GetTransmissionGlyphs();
            if(!MainFormHelper.TakeScreens) {
                bandedGridView1.ActiveFilterString = "InRange([ModelPrice], 25000, 80000) And (IsOutlookIntervalToday([SalesDate]) Or IsOutlookIntervalYesterday([SalesDate]))";
                bandedGridView1.MRUFiltersInfo.Add(new ViewFilter(bandedGridView1, CriteriaOperator.Parse("InRange([ModelPrice], 50000, 150000)")));
                bandedGridView1.MRUFiltersInfo.Add(new ViewFilter(bandedGridView1, CriteriaOperator.Parse("IsOutlookIntervalToday([SalesDate]) Or IsOutlookIntervalYesterday([SalesDate])")));
                bandedGridView1.MRUFiltersInfo.Add(new ViewFilter(bandedGridView1, CriteriaOperator.Parse("[SalesDate] > AddDays(LocalDateTimeToday(),-25)")));
            }
            else bandedGridView1.ActiveFilterString = "([ModelPrice] >= 25000 And [ModelPrice] <= 80000)";
        }
        protected override string[] WhatsThisCodeFileNames {
            get { return new string[] { "ExcelFiltering" }; }
        }
        protected override string WhatsThisXMLFileName {
            get { return "ExcelFiltering"; }
        }
        // Addding custom Filters and Data Items customization
        //<gridControl1>
        void gridView_FilterPopupExcelData(object sender, FilterPopupExcelDataEventArgs e) {
            string fieldName = e.Column.FieldName;
            if(e.Column == bcModification) {
                // Filter Items
                e.AddFilter("<image=A><nbsp>Automatic Transmission (6-speed)", "Contains([" + fieldName + "], '6A')", true);
                e.AddFilter("<image=A><nbsp>Automatic Transmission (8-speed)", "Contains([" + fieldName + "], '8A')", true);
                e.AddFilter("<image=M><nbsp>Manual Transmission (6-speed)", "Contains([" + fieldName + "], '6M')", true);
                e.AddFilter("<image=M><nbsp>Manual Transmission (7-speed)", "Contains([" + fieldName + "], '7M')", true);
                e.AddFilter("<image=V><nbsp>Variomatic Transmission", "Contains([" + fieldName + "], 'VA')", true);
                e.AddFilter("<b>Limited Edition</b>", "Contains([" + fieldName + "], 'Limited')", true);
                // Data Items
                foreach(var item in e.DataItems) {
                    if(item.Text.Contains("V6"))
                        item.HtmlText = item.Text.Replace("V6", "<b>V6</b>");
                    if(item.Text.Contains("V8"))
                        item.HtmlText = item.Text.Replace("V8", "<b>V8</b>");
                    if(item.Text.Contains("Limited"))
                        item.HtmlText = "<image=Ltd><nbsp>" + item.Text;
                }
            }
            if(e.Column == bcMPGCity) { // 12-28
                e.AddFilter("Fuel Economy (<color=@Information>High</color>)", "[" + fieldName + "]>25", true);
                e.AddFilter("Fuel Economy (<color=@Warning>Medium</color>)", "[" + fieldName + "]>=15 AND [" + fieldName + "]<=25", true);
                e.AddFilter("Fuel Economy (<color=@Critical>Low</color>)", "[" + fieldName + "]<15", true);
            }
            if(e.Column == bcMPGHighway) { // 15-36
                e.AddFilter("Fuel Economy (<color=@Information>High</color>)", "[" + fieldName + "]>20", true);
                e.AddFilter("Fuel Economy (<color=@Warning>Medium</color>)", "[" + fieldName + "]>=20 AND [" + fieldName + "]<=30", true);
                e.AddFilter("Fuel Economy (<color=@Critical>Low</color>)", "[" + fieldName + "]>20", true);
            }
            // sales from newer to older by years
            if(e.Column == bcSalesDate) {
                // using low-level API for inplace values and texts reordering
                Array.Sort(e.Values, e.DisplayTexts, YearsFromNewerToOlder.Comparer);
            }
        }
        //</gridControl1>
        #region GenerateReport
        public override bool AllowGenerateReport {
            get { return false; }
        }
        #endregion
        sealed class YearsFromNewerToOlder : IComparer<object> {
            public static IComparer<object> Comparer = new YearsFromNewerToOlder();
            YearsFromNewerToOlder() { }
            //
            int IComparer<object>.Compare(object x, object y) {
                var dateX = (DateTime)x;
                var dateY = (DateTime)y;
                var yearsResultFromNewerToOlder = dateY.Year.CompareTo(dateX.Year);
                if(yearsResultFromNewerToOlder == 0) 
                    return dateX.CompareTo(dateY);
                return yearsResultFromNewerToOlder;
            }
        }
    }
}
